### PR TITLE
Add put method for Collections

### PR DESCRIPTION
**Pull request by kjsanger**
Change defaults for checksum-related API method keywords to False in order to make behaviour and responsibility for the decision clearer. API callers are now responsible for setting these flags to True where they are needed.

Separate checksum retrieval, updates and verification more clearly.

Add a checkum keyword to DataObject::list.

Add checksum size/consistency check for zero-length data objects.

### CODE REVIEW

File: src/partisan/irods.py [hunk 1]
```diff
@@ -2438,17 +2650,98 @@ def get(self, local_path: Union[Path, str], **kwargs):
...
+            raise ValueError(f"Local path '{local_path}' is not a directory")
+
+        self.create(exist_ok=True, timeout=timeout, tries=tries)
+
+        if recurse:
+            for root, dirs, files in os.walk(local_path):
+                for d in dirs:
+                    p = Path(root, d)
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Creating collection", local_path=d, remote_path=r)
```
Comment (nerdstrike) [inline src/partisan/irods.py:2707]: f-string needless?

File: src/partisan/irods.py [hunk 2]
```diff
@@ -2438,17 +2650,98 @@ def get(self, local_path: Union[Path, str], **kwargs):
...
+                    Collection(r).create(exist_ok=True, timeout=timeout)
+                for f in files:
+                    p = Path(root, f)
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Putting data object", local_path=p, remote_path=r)
```
Comment (nerdstrike) [inline src/partisan/irods.py:2712]: Also a useless f

File: src/partisan/irods.py [hunk 3]
```diff
@@ -2438,17 +2650,98 @@ def get(self, local_path: Union[Path, str], **kwargs):
...
+                        force=force,
+                        timeout=timeout,
+                        tries=tries,
+                    )
+        else:
+            for p in Path(local_path).iterdir():
+                if p.is_dir():
+                    r = PurePath(self.path, p.relative_to(local_path))
+                    log.debug(f"Creating collection", local_path=p, remote_path=r)
+                    Collection(r).create(exist_ok=True, timeout=timeout)
```
Comment (nerdstrike) [inline src/partisan/irods.py:2729]: Does it make sense to archive an empty folder when you're not recursing? Seems a little futile.

**Review by nerdstrike — CHANGES_REQUESTED** [review]
Some minor mistakes and a couple of questions for you. Hard for me to tell if you missed any implications from changing the defaults on checksum calculation

File: tests/test_irods.py [hunk 4]
```diff
@@ -768,6 +806,97 @@ def test_make_data_object_collection_path(self, simple_collection):
...
+        local_path = Path("./tests/data/simple/data_object/lorem.txt").absolute()
+        with pytest.raises(ValueError, match="mismatch"):
+            obj.put(
+                local_path,
+                calculate_checksum=True,
+                compare_checksums=True,
+                local_checksum=lambda _: "a bad checksum",
+            )
+
+    @m.it("Raises an error if a supplied local checksum string does not match")
```
[the remark this answers is not in the document]
Comment (kjsanger) [reply tests/test_irods.py:891]: Fixed with `calculate_checksum=True`

File: src/partisan/irods.py (hunk 3 above)
Comment (kjsanger) [reply src/partisan/irods.py:2729]: It makes reasoning about what happens easier; exact mirror versus mirror, but with an exception.

File: tests/test_irods.py [hunk 5]
```diff
@@ -789,7 +918,7 @@ def test_get_data_object(self, tmp_path, simple_data_object):
         obj = DataObject(simple_data_object)
 
         local_path = tmp_path / simple_data_object.name
-        size = obj.get(local_path)
+        size = obj.get(local_path, verify_checksum=True)
```
[the remark this answers is not in the document]
Comment (kjsanger) [reply tests/test_irods.py:921]: It goes through `Baton::get` which does have this (`BatonPool::get` gets a client from the pool).

**Review by nerdstrike — APPROVED** [review]

**Pull request merged**